plugin/plugins/filesystem/nodes/base.go

```diff
@@ -0,0 +1,241 @@
+
+	This context is expected to be valid for the lifetime of the file system
+	and canceled on `Close` by references returned from `Attach` only.
+	*/
+	filesystemCtx    context.Context
+	filesystemCancel context.CancelFunc
+
+	/* The operations context should be unique to the file reference.
+	References derived from `file.Walk` should contain a new context and cancel.
+	(Likely populated with one derived from the existing fs context)
+	This context is expected to be valid as long as the file is being referenced by a particular FID
+	and should be canceled during `Close`.
+	*/
+	operationsCtx    context.Context
+	operationsCancel context.CancelFunc
+
+	/* Format the namespace as if it were a rooted directory, sans trailing slash.
+	e.g. `/ipfs`
+	The base relative path is appended to the namespace for core requests upon calling `.CorePath()`.
+	*/
+	coreNamespace string
+	core          coreiface.CoreAPI
+}
+
+func newIPFSBase(ctx context.Context, coreNamespace string, core coreiface.CoreAPI, ops ...nodeopts.AttachOption) IPFSBase {
+	options := nodeopts.AttachOps(ops...)
+	return IPFSBase{
+		Base: newBase(ops...),
+		OverlayFileMeta: OverlayFileMeta{
+			parent: options.Parent,
+		},
+		coreNamespace: coreNamespace,
+		core:          core,
+		filesystemCtx: ctx,
+	}
+}
+
+/* general helpers */
+
+func (b *Base) ninePath() p9Path { return b.qid.Path }
+func (b *Base) String() string   { return gopath.Join(b.Trail...) }
+
+func (ib *IPFSBase) String() string {
+	return gopath.Join(append([]string{ib.coreNamespace}, ib.Base.String())...)
+}
+
+// see filesystemCtx section in IPFSBase comments
+func (ib *IPFSBase) forkFilesystem() error {
+	if err := ib.filesystemCtx.Err(); err != nil {
+		return err
+	}
+	ib.filesystemCtx, ib.filesystemCancel = context.WithCancel(ib.filesystemCtx)
+	return nil
+}
+
+// see operationsCtx section in IPFSBase comments
+func (ib *IPFSBase) forkOperations() error {
+	if err := ib.filesystemCtx.Err(); err != nil {
+		return err
+	}
+	ib.operationsCtx, ib.operationsCancel = context.WithCancel(ib.filesystemCtx)
+	return nil
+}
+
+func (b *IPFSBase) callCtx() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(b.filesystemCtx, 30*time.Second)
+}
+
+func (ib *IPFSBase) CorePath(names ...string) corepath.Path {
+	return corepath.Join(rootPath(ib.coreNamespace), append(ib.Trail, names...)...)
+}
+
+/* base operation methods to build on */
+
+func (b *Base) close() error {
+	b.Logger.Debugf("closing: {%d}%q", b.qid.Path, b.String())
+	b.closed = true
+	return nil
+}
+
+func (ib *IPFSBase) close() error {
+	lastErr := ib.Base.close()
+	if lastErr != nil {
+		ib.Logger.Error(lastErr)
+	}
+
+	if ib.filesystemCancel != nil {
+		/* TODO: only do this on the last close to the root
+		if ib.proxy != nil {
+		    if err := ib.proxy.Close(); err != nil {
+				ib.Logger.Error(err)
+			}
+			lastErr = err
+		}
+		ib.filesystemCancel()
+		*/
```

Comment:
That was a mistake.
The filesystem context should have been canceled, but the proxy left open.
https://github.com/ipfs/go-ipfs/pull/6612/commits/0fa59833378c03651a49e28f2398a56a9b9659b5
un-comments the cancel, and adds in finalizers to close any proxy file systems.

A better approach here though would probably be to add a reference count that's incremented on `Attach`, and checked during `Close` so that the error value doesn't get lost on the final call to `someFs.Close`.